packages/e2e-test-utils-playwright/src/admin/visit-site-editor.ts

```diff
@@ -8,64 +8,42 @@ import { addQueryArgs } from '@wordpress/url';
  */
 import type { Admin } from './';
 
-export interface SiteEditorQueryParams {
-	postId: string | number;
-	postType: string;
+interface SiteEditorOptions {
+	postId?: string | number;
+	postType?: string;
+	path?: string;
+	canvas?: string;
+	showWelcomeGuide?: boolean;
 }
 
-const CANVAS_SELECTOR = 'iframe[title="Editor canvas"i] >> visible=true';
-
 /**
- * Visits the Site Editor main page
- *
- * By default, it also skips the welcome guide. The option can be disabled if need be.
+ * Visits the Site Editor main page.
  *
  * @param this
- * @param query            Query params to be serialized as query portion of URL.
- * @param skipWelcomeGuide Whether to skip the welcome guide as part of the navigation.
+ * @param options Options to visit the site editor.
  */
 export async function visitSiteEditor(
 	this: Admin,
-	query: SiteEditorQueryParams,
-	skipWelcomeGuide = true
+	options: SiteEditorOptions = {}
 ) {
-	const path = addQueryArgs( '', {
-		...query,
+	const query = addQueryArgs( '', {
+		postId: options.postId,
+		postType: options.postType,
+		path: options.path,
+		canvas: options.canvas,
 	} ).slice( 1 );
 
-	await this.visitAdminPage( 'site-editor.php', path );
-
-	if ( skipWelcomeGuide ) {
-		await this.page.evaluate( () => {
-			window.wp.data
-				.dispatch( 'core/preferences' )
-				.set( 'core/edit-site', 'welcomeGuide', false );
-
-			window.wp.data
-				.dispatch( 'core/preferences' )
-				.set( 'core/edit-site', 'welcomeGuideStyles', false );
+	await this.visitAdminPage( 'site-editor.php', query );
 
-			window.wp.data
-				.dispatch( 'core/preferences' )
-				.set( 'core/edit-site', 'welcomeGuidePage', false );
-
-			window.wp.data
-				.dispatch( 'core/preferences' )
-				.set( 'core/edit-site', 'welcomeGuideTemplate', false );
+	if ( ! options.showWelcomeGuide ) {
+		await this.editor.setPreferences( 'core/edit-site', {
+			welcomeGuide: false,
+			welcomeGuideStyles: false,
+			welcomeGuidePage: false,
+			welcomeGuideTemplate: false,
 		} );
 	}
 
-	// Check if the current page has an editor canvas first.
-	if ( ( await this.page.locator( CANVAS_SELECTOR ).count() ) > 0 ) {
-		// The site editor initially loads with an empty body,
-		// we need to wait for the editor canvas to be rendered.
-		await this.page
-			.frameLocator( CANVAS_SELECTOR )
-			.locator( 'body > *' )
-			.first()
-			.waitFor();
-	}
```

Comment:
This code removal doesn't look related to the changes proposed in this PR. As discussed in #54911, there were usually good reasons for introducing similar guardrails.

Maybe we should split this into a separate PR. What do you think?

Reply:
Ugh, nice catch! I actually forgot to restore this part. I wasn't planning on removing it in this PR, only running a single round without it to see how it goes. Restoring it now.

Reply:
It's good, though, that now we know only perf tests fail without it - these tests require specific handling, so it might be good to move this safeguard there if it's not necessary here for the _regular_ E2Es. Anyway, it's just a note for myself, I guess - let's leave that for another PR.

Reply:
Restored in https://github.com/WordPress/gutenberg/pull/55099/commits/9559d8e2e1981a7d9234a793a9fcb74f97d3ee4a

Reply:
I had to remove it as the slight timing change caused by using `setPreferences` in `visitSiteEditor` revealed that [the part above](https://github.com/WordPress/gutenberg/pull/55099#pullrequestreview-1673128184) is actually faulty. Before, it checked the canvas iframe(s) **before** they were attached, so that check has always been skipped and moved instantly to wait until the loading spinner is hidden. Now, because `setPreferences` [adds a little delay](https://github.com/WordPress/gutenberg/blob/34f93b329f0c23718817febc7fbfca2b8b47b26b/packages/e2e-test-utils-playwright/src/editor/set-preferences.ts#L23) by waiting until the `window.wp.data` is available, all the editor iframes get attached in the meantime, which ends u breaking the locator strict mode violation:

```
at /home/runner/work/gutenberg/gutenberg/test/e2e/specs/site-editor/patterns.spec.js:165:3
Error:   1) [chromium] › site-editor/patterns.spec.js:137:2 › Patterns › search and filter patterns ───────

    Retry #1 ───────────────────────────────────────────────────────────────────────────────────────
    Error: locator.waitFor: Error: strict mode violation: locator('iframe[title="Editor canvas"i]').locator('visible=true') resolved to 3 elements:
        1) <iframe tabindex="-1" aria-hidden="true" title="Editor c…></iframe> aka locator('#edit-site-patterns-synced-footer')
        2) <iframe tabindex="-1" aria-hidden="true" title="Editor c…></iframe> aka locator('#edit-site-patterns-unsynced-header')
        3) <iframe tabindex="-1" aria-hidden="true" title="Editor c…></iframe> aka locator('#edit-site-patterns-unsynced-footer')
```

ℹ️ The other iframes are previews of the patterns created in that test.